### ISSUE TITLE
LazyInitializationException while fetching entity with lazy associated field

### ISSUE THREAD

**Issue by DnEgorWeb** [quarkusio/quarkus#29189]
### Describe the bug

Here is [SO question](https://stackoverflow.com/questions/74391462/how-to-fetch-lazy-associations-in-hibernate-reactive) I've created. @DavideD had a look and said it might be a bug. My Quarkus project uses kotlin/hibernate reactive. When fetching entity with nested lazy association Hibernate throws `LazyInitializationException` despite fetching the association with Mutiny.fetch. I'm not sure if it's a bug, but at least error message a little misleading and I can't figure out what's wrong here.

### Expected behavior

The entity successfully fetched including associated entities.

### Actual behavior

`LazyInitializationException` is thrown.

### How to Reproduce?

[I've created demo project, the readme file has steps to reproduce](https://github.com/DnEgorWeb/HibernateReactiveLazyFetch)
Basically you just need to run the project and execute 3 curl requests.

### Output of `uname -a` or `ver`

Darwin MacBook-Pro-2.local 22.1.0 Darwin Kernel Version 22.1.0: Sun Oct  9 20:14:54 PDT 2022; root:xnu-8792.41.9~2/RELEASE_X86_64 x86_64

### Output of `java -version`

openjdk version "19.0.1" 2022-10-18

### Quarkus version or git rev

latest LTS

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Gradle 7.5.1

### Additional information

The issue happens only for one-to-many / many-to-one relationship (does not manner what entity you're trying to fetch - Book or Page, same error). Many-to-many works fine.

**Comment by DavideD** [comment]
The problem is that the default equals/hashcode from the entities includes the associations.
So, when you run the first select query, if they are called before you've fetched the association, it will cause the error.

You need to change your entities to something like:
```
@Entity(name = "page")
internal data class Page(
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "id")
    val pageId: Long? = null,

    @ManyToOne(fetch = FetchType.LAZY)
    @Fetch(FetchMode.JOIN)
    @JsonIgnore
    var book: Book? = null,
){
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as Page

        if (pageId != other.pageId) return false

        return true
    }

    override fun hashCode(): Int {
        return pageId?.hashCode() ?: 0
    }
}

package org.acme

import org.hibernate.annotations.Fetch
import org.hibernate.annotations.FetchMode
import javax.persistence.*

@Entity(name = "book")
internal data class Book (
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "id")
    val bookId: Long? = null,

    @OneToMany(mappedBy = "book", fetch = FetchType.LAZY)
    @Fetch(FetchMode.JOIN)
    var pages: MutableSet<Page> = mutableSetOf(),
) {
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as Book

        if (bookId != other.bookId) return false

        return true
    }

    override fun hashCode(): Int {
        return bookId?.hashCode() ?: 0
    }
}
```

Normally, you shouldn't include the identifier, but there are no other fields this simple example.
The Hibernate ORM documentation has a paragraph about [how to implement hashcode and equals for entities](https://docs.jboss.org/hibernate/orm/6.1/userguide/html_single/Hibernate_User_Guide.html#mapping-model-pojo-equalshashcode) that should give you a better explanation.

Another couple of things I've noticed:
* I don't think you need to add `@RegisterForReflection`. That's only helpful in specific circumstances and this shouldn't be one of them.
* Many to one associations are usually initialized with null. That's because you might want to check if the association exists and usually, you will do it with a null check

**Comment by DnEgorWeb** [comment]
Thank you for detailed explanation, Davide.
